crates/wasi/src/preview2/table.rs

```diff
@@ -87,23 +112,89 @@ impl Table {
         Ok(Resource::new_own(idx))
     }
 
-    fn push_(&mut self, e: TableEntry) -> Result<u32, TableError> {
-        // NOTE: The performance of this new key calculation could be very bad once keys wrap
-        // around.
-        if self.map.len() == u32::MAX as usize {
-            return Err(TableError::Full);
+    /// Pop an index off of the free queue, if it's not empty.
+    fn pop_free_queue_(&mut self) -> Option<usize> {
+        let res = self.free_head;
+        if let Some(ix) = res {
+            // Advance free_head to the next entry if one is available.
+            match &self.entries[ix] {
+                Entry::Free { next: Some(next) } => {
+                    self.free_head.replace(*next);
+                }
+
+                Entry::Free { next: None } => {
+                    self.free_head = None;
+                    self.free_tail = None;
+                }
+
+                Entry::Occupied { .. } => unreachable!(),
+            }
         }
-        loop {
-            let key = self.next_key;
-            self.next_key = self.next_key.wrapping_add(1);
-            if self.map.contains_key(&key) {
-                continue;
+
+        res
+    }
+
+    fn free_entry_(&mut self, ix: usize) -> TableEntry {
+        if let Some(tail) = self.free_tail {
+            debug_assert!(self.free_head.is_some());
+            let tail = &mut self.entries[tail];
+            match tail {
+                Entry::Free { next } => {
+                    debug_assert!(next.is_none());
+                    next.replace(ix);
+                }
+
+                Entry::Occupied { .. } => unreachable!(),
             }
-            self.map.insert(key, e);
-            return Ok(key);
+        } else {
+            debug_assert!(self.free_head.is_none());
+            self.free_head.replace(ix);
+        }
+
+        let entry = match std::mem::replace(&mut self.entries[ix], Entry::Free { next: None }) {
+            Entry::Occupied { entry } => entry,
+            Entry::Free { .. } => unreachable!(),
+        };
+
+        self.free_tail = Some(ix);
+
+        entry
+    }
+
+    /// Push a new entry into the table, returning its handle. The following strategy is used:
+    /// 1. if there's capacity in the vector, push a new entry at the end,
+    /// 2. if the free queue is non-empty, use the entry at the head of the queue,
+    /// 3. finally, grow the vector and push on the end.
+    fn push_(&mut self, e: TableEntry) -> Result<u32, TableError> {
+        let len = self.entries.len();
+        if len < self.entries.capacity() {
+            let ix = len.try_into().map_err(|_| TableError::Full)?;
+            self.entries.push(Entry::Occupied { entry: e });
+            Ok(ix)
```

Comment:
I was naively expecting a slab-like allocator when taking a look at this, which wouldn't have this branch. Could this perhaps be removed to help simplify this a bit further?